spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/couchbase/embedded/EmbeddedCouchbaseAutoConfiguration.java

```diff
@@ -0,0 +1,87 @@
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.boot.autoconfigure.couchbase.embedded;
+
+import java.io.IOException;
+import java.util.Collections;
+
+import com.couchbase.client.java.CouchbaseBucket;
+import com.couchbase.mock.Bucket;
+import com.couchbase.mock.BucketConfiguration;
+import com.couchbase.mock.CouchbaseMock;
+
+import org.springframework.boot.autoconfigure.AutoConfigureBefore;
+import org.springframework.boot.autoconfigure.EnableAutoConfiguration;
+import org.springframework.boot.autoconfigure.condition.ConditionalOnBean;
+import org.springframework.boot.autoconfigure.condition.ConditionalOnClass;
+import org.springframework.boot.autoconfigure.condition.ConditionalOnMissingBean;
+import org.springframework.boot.autoconfigure.couchbase.CouchbaseAutoConfiguration;
+import org.springframework.boot.autoconfigure.couchbase.CouchbaseProperties;
+import org.springframework.boot.context.properties.EnableConfigurationProperties;
+import org.springframework.context.annotation.Bean;
+import org.springframework.context.annotation.Configuration;
+
+/**
+ * {@link EnableAutoConfiguration Auto-configuration} for Embedded Couchbase.
+ *
+ * @author Alex Derkach
+ * @since 2.0.0
+ */
+@Configuration
+@EnableConfigurationProperties({CouchbaseProperties.class})
+@AutoConfigureBefore(CouchbaseAutoConfiguration.class)
+@ConditionalOnClass({CouchbaseBucket.class, CouchbaseMock.class})
+public class EmbeddedCouchbaseAutoConfiguration {
+
+	@Bean(destroyMethod = "stop")
+	@ConditionalOnMissingBean
+	public CouchbaseMock couchbaseMock(CouchbaseProperties properties) throws IOException, InterruptedException {
+
+		CouchbaseProperties.Bucket bucketProperties = properties.getBucket();
+
+		BucketConfiguration bucketConfiguration = new BucketConfiguration();
+		bucketConfiguration.name = bucketProperties.getName();
+		bucketConfiguration.password = bucketProperties.getPassword();
+		bucketConfiguration.type = Bucket.BucketType.COUCHBASE;
+		bucketConfiguration.numReplicas = 1;
+		bucketConfiguration.numNodes = 1;
+
+		CouchbaseMock couchbaseMock = new CouchbaseMock(0, Collections.singletonList(bucketConfiguration));
+		couchbaseMock.start();
+		couchbaseMock.waitForStartup();
+
+		String host = couchbaseMock.getHttpHost();
+		int bootstrapHttpDirectPort = couchbaseMock.getHttpPort();
+		int bootstrapCarrierDirectPort = couchbaseMock.getBuckets().get(properties.getBucket().getName()).getServers()[0].getPort();
+
+		couchbaseMock.startHarakiriMonitor(host + ":" + bootstrapHttpDirectPort, false);
+
+		properties.setBootstrapHosts(Collections.singletonList(host));
```

Comment:
Thanks it's much better but that's not what I have in mind with "make that its own class with the necessary parameters that you need". This change is better as you're not relying on the order of processing of `CouchbaseProperties` but you're still changing its state and it is a singleton.

What I have more in mind is to apply those customizations to your extension of `SpringBootConfiguration`, either via a protected method of some sort or any other suitable means.